### PR TITLE
Town Grow Functions

### PR DESCRIPTION
**Pull request by duncanspumpkin**
This still has my debug code in it but that can be deleted before merge. I can't think of a good name for these functions just yet but since that has been my excuse for not putting them up for review for 2 months might as well just leave the names as is. I verified these with the replays a bit but could probably be rechecked.

### CODE REVIEW

File: src/OpenLoco/src/Town.cpp [hunk 1]
```diff
@@ -204,4 +210,137 @@ namespace OpenLoco
         }
         return StringIds::town_size_hamlet;
     }
+
+    // 0x00463BD2
+    template<typename Func>
+    static void sub_463BD2(const World::Pos2& topLeftLoc, uint8_t searchSize, Func&& predicate)
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Town.cpp:216]: I think we can name this to something like `predicateSearch`? Up to you if you want to leave it for #1742 to tackle.

File: src/OpenLoco/src/Town.cpp [hunk 2]
```diff
@@ -204,4 +210,137 @@ namespace OpenLoco
...
+        }
+    }
+
+    // 0x00497FFC
+    std::optional<Sub497FFCResult> Town::sub_497FFC()
+    {
+        registers regs;
+        regs.esi = X86Pointer(this);
+        call(0x00497FFC, regs);
+        std::optional<Sub497FFCResult> origRes;
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Town.cpp:263]: Shouldn't this be removed prior to merging?

File: src/OpenLoco/src/Town.h [hunk 3]
```diff
@@ -29,6 +30,13 @@ namespace OpenLoco
         metropolis,
     };
 
+    struct Sub497FFCResult
+    {
+        World::Pos3 roadStart;
+        uint16_t tad;
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Town.h:36]: tad?

**Review by AaronVanGeffen — CHANGES_REQUESTED** [review]
This seems to have changed behaviour somewhat. I tested town growth using sandbox mode, picking the same town in Oil Oasis on both this PR and a master build for a baseline. The build for this PR seems to be building/upgrading more quickly. Notably more of the same building seems to be used, turning many buildings into skyscrapers after clicking the grow button a few times.

Can you reproduce this issue? Do replay tests pass, long-term?

File: src/OpenLoco/src/Town.h (hunk 3 above)
Comment (duncanspumpkin) [reply src/OpenLoco/src/Town.h:36]: track and direction. I use tad in a few places.

**Comment by duncanspumpkin** [comment]
> This seems to have changed behaviour somewhat. I tested town growth using sandbox mode, picking the same town in Oil Oasis on both this PR and a master build for a baseline. The build for this PR seems to be building/upgrading more quickly. Notably more of the same building seems to be used, turning many buildings into skyscrapers after clicking the grow button a few times.
> 
> Can you reproduce this issue? Do replay tests pass, long-term?

Hmm that shouldn't be possible the whole reason behind the asserts is that no behaviour has changed. I'll look into it again

File: src/OpenLoco/src/Town.cpp (hunk 1 above)
Comment (duncanspumpkin) [reply src/OpenLoco/src/Town.cpp]: I was thinking maybe boxSearch or squareSearch.

File: src/OpenLoco/src/Town.h (hunk 3 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/src/Town.h]: Would it be possible to use a type for this to make it more intuitive?

Comment (duncanspumpkin) [reply src/OpenLoco/src/Town.h]: i do actually have a type for it but annoyingly there are like 4-5 variations of it. I can't decide the best approach.

File: src/OpenLoco/src/Town.cpp (hunk 1 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/src/Town.cpp]: Per the ASCII art, I'm thinking `squareSearch` takes the `topLeftLoc` as a corner of the search square. This is reasonably intuitive. Let's go with that then.

Comment (duncanspumpkin) [reply src/OpenLoco/src/Town.cpp]: So turns out my ascii art was wrong. Its still square but i made a mistake in the drawing.

**Review by AaronVanGeffen — APPROVED** [review]

**Pull request merged**